lib/lf/uscalfe/lagr_fe.h

```diff
@@ -1155,6 +1155,500 @@ const Eigen::MatrixXi FeLagrangeO2Quad<SCALAR>::ksegment_to_quad_mapping_ =
                              2, 2).finished();
 // clang-format on
 
+/**
+ * @headerfile lf/uscalfe/uscalfe.h
+ * @brief Cubic Lagrangian finite elment on a triangular reference element
+ *
+ * This is a specialization of ScalarReferenceFiniteElement.
+ * Refer to its documentation.
+ *
+ * The reference shape functions are combinations of the barycentric coordinate
+ * functions on the reference triangle.
+ *
+ * The first three shape functions are associated with vertices,
+ * the next six with edges of the triangle. The last shape function is an
+ * interior shape function.
+ *
+ * @see ScalarReferenceFiniteElement
+ */
+template <typename SCALAR>
+class FeLagrangeO3Tria final : public ScalarReferenceFiniteElement<SCALAR> {
+ public:
+  FeLagrangeO3Tria(const FeLagrangeO3Tria&) = default;
+  FeLagrangeO3Tria(FeLagrangeO3Tria&&) noexcept = default;
+  FeLagrangeO3Tria& operator=(const FeLagrangeO3Tria&) = default;
+  FeLagrangeO3Tria& operator=(FeLagrangeO3Tria&&) noexcept = default;
+  FeLagrangeO3Tria() = default;
+  ~FeLagrangeO3Tria() override = default;
+
+  [[nodiscard]] base::RefEl RefEl() const override {
+    return base::RefEl::kTria();
+  }
+
+  [[nodiscard]] unsigned Degree() const override { return 3; }
+
+  [[nodiscard]] size_type NumRefShapeFunctions() const override { return 10; }
+
+  /**
+   * @brief One shape function attached to each node
+   * and two to each edge of the triangle. There is one interior
+   * shape function.
+   * @copydoc ScalarReferenceFiniteElement::NumRefShapeFunctions(dim_t)
+   */
+  [[nodiscard]] size_type NumRefShapeFunctions(dim_t codim) const override {
+    LF_ASSERT_MSG(codim <= 2, "Illegal codim" << codim);
+    return (codim == 1) ? 2 : 1;
+  }
+
+  /**
+   * @brief One shape function attached to each node
+   * and two to each edge of the triangle. There is one interior
+   * shape function.
+   * @copydoc
+   * ScalarReferenceFiniteElement::NumRefShapeFunctions(dim_t,sub_idx_t)
+   */
+  [[nodiscard]] size_type NumRefShapeFunctions(
+      dim_t codim, sub_idx_t /*subidx*/) const override {
+    LF_ASSERT_MSG(codim <= 2, "Illegal codim" << codim);
+    return (codim == 1) ? 2 : 1;
+  }
+
+  [[nodiscard]] Eigen::Matrix<SCALAR, Eigen::Dynamic, Eigen::Dynamic>
+  EvalReferenceShapeFunctions(const Eigen::MatrixXd& refcoords) const override {
+    LF_ASSERT_MSG(refcoords.rows() == 2,
+                  "Reference coordinates must be 2-vectors");
+
+    const size_type n_pts(refcoords.cols());
+    Eigen::Matrix<SCALAR, Eigen::Dynamic, Eigen::Dynamic> result(10, n_pts);
+
+    // evaluation of the barycentric coordinate functions
+    auto lambda0 = 1 - refcoords.row(0).array() - refcoords.row(1).array();
```

Comment:
It just takes replacing `auto` with `Eigen::Array<double,1,Eigen::Dynamic>`.

Reply:
However, note that this will involve dynamics memory allocation, which is much more expensive that any memory-local computations. So the performance gain is not predictable.

Reply:
I switched this to `Eigen::Array<double,1,Eigen::Dynamic>`. However I did not perform any experiments to check  the performance impact.

Reply:
The type of `1 - refcoords.row(0).array() - refcoords.row(1).array().eval();` should be exactly `Eigen::Array<double, 1, Eigen::Dynamic>` so replacing auto with `Eigen::Array<double, 1, Eigen::Dynamic>` doesn't affect performance. But it specifies the type exactly which can be nicer to read...
Unfortunately we cannot avoid the heap allocation easily here. In my experience the heap allocations that we have to do here are not cheap, I remember that we got a siginifacnt speedup when we go rid of them (HyDi as well as NgSolve have their own customized "heap allocator" which works essentially like a stack).

Reply:
We can of course avoid one heap allocation if we don't make the change I proposed. I guess this is what @hiptmair meant ;) But considering that `lambda` incurs two additions everytime on a whole array, and seeing that lambda is used 30 times later on, it probably makes sense. I.e. I think `time(allocation) < 30*(two additions)*lambda0.size()`. But if you really want to know it, measure it!